src/actions/sources/loadSourceText.js

```diff
@@ -25,29 +28,51 @@ async function loadSource(source: Source, { sourceMaps, client }) {
   };
 }
 
+function defer() {
+  let resolve = () => {};
+  let reject = () => {};
+  const promise = new Promise((_res, _rej) => {
+    resolve = _res;
+    reject = _rej;
+  });
+
+  return { resolve, reject, promise };
+}
+
 /**
  * @memberof actions/sources
  * @static
  */
 export function loadSourceText(source: Source) {
   return async ({ dispatch, getState, client, sourceMaps }: ThunkArgs) => {
+    const deferred = defer();
+
+    deferred.resolve();
     // Fetch the source text only once.
-    if (source.text) {
+    if (isLoaded(source)) {
       return Promise.resolve(source);
     }
 
+    if (isLoading(source) || requests.has(source.id)) {
+      return requests.get(source.id);
+    }
+
+    requests.set(source.id, deferred.promise);
     await dispatch({
       type: "LOAD_SOURCE_TEXT",
       source: source,
       [PROMISE]: loadSource(source, { sourceMaps, client })
     });
 
+    requests.delete(source.id);
+
     const newSource = getSource(getState(), source.id).toJS();
     if (newSource.isWasm) {
       return;
     }
 
     await setSource(newSource);
     dispatch(setSymbols(source.id));
+    return deferred.resolve();
```

Comment:
i dont think that this is doing more than the async function, we might not need it